pimd/pim6_cmd.c

```diff
@@ -0,0 +1,447 @@
+ */
+
+#include <zebra.h>
+
+#include "lib/json.h"
+#include "command.h"
+#include "if.h"
+#include "prefix.h"
+#include "zclient.h"
+#include "plist.h"
+#include "hash.h"
+#include "nexthop.h"
+#include "vrf.h"
+#include "ferr.h"
+#include "lib/northbound_cli.h"
+
+#include "pimd.h"
+#include "pim6_cmd.h"
+#include "pim_vty.h"
+#include "pim_errors.h"
+#include "pim_nb.h"
+
+#ifndef VTYSH_EXTRACT_PL
+#include "pimd/pim6_cmd_clippy.c"
+#endif
+
+DEFPY (interface_ipv6_pim,
+       interface_ipv6_pim_cmd,
+       "ipv6 pim",
+       IPV6_STR
+       PIM_STR)
+{
+	nb_cli_enqueue_change(vty, "./pim-enable", NB_OP_MODIFY, "true");
+
+	return nb_cli_apply_changes(vty,
+			FRR_PIM_INTERFACE_XPATH,
+			"frr-routing:ipv6");
+
+}
+
+DEFPY (interface_no_ipv6_pim,
+       interface_no_ipv6_pim_cmd,
+       "no ipv6 pim",
+       NO_STR
+       IPV6_STR
+       PIM_STR)
+{
+	const struct lyd_node *mld_enable_dnode;
+	char mld_if_xpath[XPATH_MAXLEN];
+
+	int printed =
+		snprintf(mld_if_xpath, sizeof(mld_if_xpath),
+			 "%s/frr-gmp:gmp/address-family[address-family='%s']",
+			 VTY_CURR_XPATH, "frr-routing:ipv6");
+
+	if (printed >= (int)(sizeof(mld_if_xpath))) {
+		vty_out(vty, "Xpath too long (%d > %u)", printed + 1,
+			XPATH_MAXLEN);
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	if (!mld_enable_dnode) {
+		nb_cli_enqueue_change(vty, mld_if_xpath, NB_OP_DESTROY, NULL);
+		nb_cli_enqueue_change(vty, ".", NB_OP_DESTROY, NULL);
+	} else {
+		if (!yang_dnode_get_bool(mld_enable_dnode, ".")) {
+			nb_cli_enqueue_change(vty, mld_if_xpath, NB_OP_DESTROY,
+					NULL);
+			nb_cli_enqueue_change(vty, ".", NB_OP_DESTROY, NULL);
+		} else
+			nb_cli_enqueue_change(vty, "./pim-enable", NB_OP_MODIFY,
+					"false");
+	}
+
+	return nb_cli_apply_changes(vty,
+			FRR_PIM_INTERFACE_XPATH,
+			"frr-routing:ipv6");
+}
+
+DEFPY (interface_ipv6_pim_drprio,
+       interface_ipv6_pim_drprio_cmd,
+       "ipv6 pim drpriority (1-4294967295)",
+       IPV6_STR
+       PIM_STR
+       "Set the Designated Router Election Priority\n"
+       "Value of the new DR Priority\n")
+{
+	nb_cli_enqueue_change(vty, "./dr-priority", NB_OP_MODIFY,
+			      drpriority_str);
+
+	return nb_cli_apply_changes(vty, FRR_PIM_INTERFACE_XPATH,
+			"frr-routing:ipv6");
+}
+
+DEFPY (interface_no_ipv6_pim_drprio,
+       interface_no_ipv6_pim_drprio_cmd,
+       "no ip pim drpriority [(1-4294967295)]",
+       NO_STR
+       IPV6_STR
+       PIM_STR
+       "Revert the Designated Router Priority to default\n"
+       "Old Value of the Priority\n")
+{
+	nb_cli_enqueue_change(vty, "./dr-priority", NB_OP_DESTROY, NULL);
+
+	return nb_cli_apply_changes(vty, FRR_PIM_INTERFACE_XPATH,
+				    "frr-routing:ipv6");
+}
+
+DEFPY (interface_ipv6_pim_hello,
+       interface_ipv6_pim_hello_cmd,
+       "ipv6 pim hello (1-65535) [(1-65535)]$hold",
+       IPV6_STR
+       PIM_STR
+       IFACE_PIM_HELLO_STR
+       IFACE_PIM_HELLO_TIME_STR
+       IFACE_PIM_HELLO_HOLD_STR)
+{
+	const struct lyd_node *mld_enable_dnode;
```

Comment:
This code looks almost same to similar V4 cmd (interface_ip_pim_hello_cmd) except the address-family (frr-routing:ipv6). This comment is common for other PIMv6 CLI cmds as well. Added similar comment in MLD & PIM CLI cmds PR as well (https://github.com/FRRouting/frr/pull/10365 & https://github.com/FRRouting/frr/pull/10387). I think, we should take a look at PIM/MLD CLI PRs as well.

I think, we should introduce a common file between V4 & V6 and move these code to that file and we can use "#if PIM_IPV == 4" to encode the address-family properly in that common file. That common abstracted function can be called in V4 (pim_cmd.c) & V6 (pim6_cmd.c).

That will help us maintain common code for both V4 & V6 and any common fixes will be go together. This comment is something similar to @donaldsharp 's suggestion on Socket-changes

Reply:
If we keep same DEFUN function for v4 and v6 ex: ip[v6] pim, then it will go to both the pim and pimv6 daemons.
Thats the reason we are keeping separate DEFUN functions, so that it will be called for specific daemon.

Reply:
Added a detailed comment with e.g. in other PIM PR.
Here is the link : https://github.com/FRRouting/frr/pull/10387/files#r801060701